File: 백준/Gold/16928. 뱀과 사다리 게임/뱀과 사다리 게임.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    static int[][][] map;
    static boolean[][] visited;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());


        //행, 열, 사다리 또는 뱀 여부, 이동 칸 값
        map = new int[10][10][2];
        visited = new boolean[10][10];
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        for (int n = 0; n < N; n++) {
            st = new StringTokenizer(br.readLine());
            int x = Integer.parseInt(st.nextToken())-1;
            int y = Integer.parseInt(st.nextToken());

            map[x / 10][(x) % 10 ][0] = 1; //사다리
            map[x / 10][(x) % 10 ][1] = y;
        }

        for (int m = 0; m < M; m++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken())-1;
            int v = Integer.parseInt(st.nextToken());

            map[u / 10][(u) % 10 ][0] = 2; //뱀
            map[u / 10][(u) % 10 ][1] = v;
        }


        int rank = bfs();
        System.out.println(rank);
    }

    public static int bfs() {
        Queue<int[]> queue = new LinkedList<>();

//        int now = 1;
        queue.add(new int[]{1, 0});
        visited[0][0] = true;

        while (!queue.isEmpty()) {
            int[] tmp = queue.poll();

            if (tmp[0] == 100) return tmp[1];
            for (int n = 1; n <= 6; n++) {
                int next = tmp[0] + n;

                int du = (next-1)/10;
                int dv = (next-1) % 10;

                if (next <= 100 && !visited[du][dv]) {
                    if (map[du][dv][0] == 0) {
                        queue.add(new int[]{next, tmp[1] + 1});
                        visited[du][dv] = true;
                    } else if (map[du][dv][0] == 1) {
                        queue.add(new int[]{map[du][dv][1], tmp[1] + 1});
                        int ladder = map[du][dv][1] - 1;
                        visited[du][dv] = true;

//                        visited[ladder / 10][(ladder) % 10] = true;


                    }else if(map[du][dv][0]==2){
                        queue.add(new int[]{map[du][dv][1], tmp[1]+1});
                        visited[du][dv] = true;

                    }

                }
            }
        }
        return -1;
    }


}

File: 백준/Gold/16928. 뱀과 사다리 게임/뱀과 사다리 게임.py
"""
주사위 수 num
현재 칸 i
i+num>100이면 이동불가
i+num 이 사다리면 사다리 타고 이동
i+num 이 뱀이면 뱀 타고 이동

#사다리와 뱀 정보가 주어질 때 100까지 도착하기 위한 주사위 횟수의 최솟값은?

#bfs
100부터 출발해서 1,2,3,4,5,6칸으로 확장.
그 칸에서 사다리 있으면 이전 칸으로 / 뱀 있으면 이후 칸으로 이동해보며
1이 가장 먼저 나올 때의 rank가 답

"""
from collections import deque


def bfs():
    q = deque([(1, 0)])
    visited = [0] * 101
    visited[1] = 1
    while q:
        now, rank = q.popleft()
        if now == 100:
            return rank
        for i in range(1, 7):
            if now + i > 100:
                break
            next = now + i
            if visited[next]:
                continue
            while ladder[next] != 0 or snake[next] != 0:
                if ladder[next] != 0:
                    next = ladder[next]
                elif snake[next] != 0:
                    next = snake[next]
            q.append((next, rank + 1))
            visited[next] =1
    return -1


N, M = map(int, input().split())
ladder = [0] * 101
snake = [0] * 101
for i in range(N):
    a, b = map(int, input().split())
    ladder[a] = b  # a에서 온 사다리
for i in range(M):
    a, b = map(int, input().split())
    snake[a] = b  # a에서 온 뱀

answer = bfs()
print(answer)